compiler/noirc_frontend/src/elaborator/types.rs

```diff
@@ -0,0 +1,1392 @@
+    fn try_remove_implicit_dereference(&mut self, object: ExprId) -> Option<ExprId> {
+        match self.interner.expression(&object) {
+            HirExpression::MemberAccess(mut access) => {
+                let new_lhs = self.try_remove_implicit_dereference(access.lhs)?;
+                access.lhs = new_lhs;
+                access.is_offset = true;
+
+                // `object` will have a different type now, which will be filled in
+                // later when type checking the method call as a function call.
+                self.interner.replace_expr(&object, HirExpression::MemberAccess(access));
+                Some(object)
+            }
+            HirExpression::Prefix(prefix) => match prefix.operator {
+                // Found a dereference we can remove. Now just replace it with its rhs to remove it.
+                UnaryOp::Dereference { implicitly_added: true } => Some(prefix.rhs),
+                _ => None,
+            },
+            _ => None,
+        }
+    }
+
+    fn bind_function_type_impl(
+        &mut self,
+        fn_params: &[Type],
+        fn_ret: &Type,
+        callsite_args: &[(Type, ExprId, Span)],
+        span: Span,
+    ) -> Type {
+        if fn_params.len() != callsite_args.len() {
+            self.push_err(TypeCheckError::ParameterCountMismatch {
+                expected: fn_params.len(),
+                found: callsite_args.len(),
+                span,
+            });
+            return Type::Error;
+        }
+
+        for (param, (arg, _, arg_span)) in fn_params.iter().zip(callsite_args) {
+            self.unify(arg, param, || TypeCheckError::TypeMismatch {
+                expected_typ: param.to_string(),
+                expr_typ: arg.to_string(),
+                expr_span: *arg_span,
+            });
+        }
+
+        fn_ret.clone()
+    }
+
+    pub(super) fn bind_function_type(
+        &mut self,
+        function: Type,
+        args: Vec<(Type, ExprId, Span)>,
+        span: Span,
+    ) -> Type {
+        // Could do a single unification for the entire function type, but matching beforehand
+        // lets us issue a more precise error on the individual argument that fails to type check.
+        match function {
+            Type::TypeVariable(binding, TypeVariableKind::Normal) => {
+                if let TypeBinding::Bound(typ) = &*binding.borrow() {
+                    return self.bind_function_type(typ.clone(), args, span);
+                }
+
+                let ret = self.interner.next_type_variable();
+                let args = vecmap(args, |(arg, _, _)| arg);
+                let env_type = self.interner.next_type_variable();
+                let expected = Type::Function(args, Box::new(ret.clone()), Box::new(env_type));
+
+                if let Err(error) = binding.try_bind(expected, span) {
+                    self.push_err(error);
+                }
+                ret
+            }
+            // ignoring env for subtype on purpose
```

Comment:
`on purpose` i.e. because the subtype is expected to have an empty `env`?

Reply:
on purpose because the arguments of a function call never place any type constraint on a closure's environment. I'll clarify the comment.